### ISSUE TITLE
Merge with IndieWeb plugin.

### ISSUE THREAD

**Issue by miklb** [indieweb/wordpress-uf2#29]
I know just moved this to the indieweb repo and added to the wp.org repo, but since this plugin can co-exist with themes, what would be the harm in merging the two?

That way, as soon as you activate the IndieWeb WordPress plugin you have basic mf2 support, and one less step. 

That may though be part of the discussion as to having the IW plugin auto-activating several plugins, in which case this plugin could be one, but kept separate.

**Comment by pfefferle** [comment]
I like the last paragraph, but I do not know how difficult this is.

To your main question: I have a strong opinion on this topic. I take the 3. and 4. topic of the IndieWeb principles very serious:

1. Build tools for yourself, not for all of your friends.
2. Eat your own dogfood.

I have very little spare time, so I concentrate on plugins, that fits my needs and I try to ignore things I do not need/actively use. So, the smaller a plugins use case is, the better it matches my and your needs. I had several problems with some features of the IndieWeb plugin, because I implemented most of the features (res-me links, author h-card) which cases some redundancy errors: https://github.com/indieweb/wordpress-indieweb/issues/40 So I was not able to use the plugin for quite some time.

I for example like the idea of the `u-featured` plugin, but if it's part of the IndieWeb plugin, I might not be able to use it.

So I would prefer small and independent but interoperable plugins and would vote against merging it.

**Comment by miklb** [comment]
I appreciate your stand for dogfooding and guiding tools for yourself. I also think IndieWeb is reaching a tipping point for gen 2 & 3 users, and WordPress is one of the more logical places for those users to migrate to, thus the idea that it is time to start looking at ways to help in that regard.

It is obvious that the WordPress IndieWeb plugins aren't going to be merged into one monolithic plugin as desired by some outside the IW WP ecosystem, or some new to IW WP as users. But if we could minimize the steps to get up and running by auto-installing the required/core plugins (with option to turn off of course) we could open up the user base.

I also realize that it's a chicken and egg situation. Without any new contributors, but expanding users, how will that affect the few current volunteer developers?

I mostly raised this question of [auto-install](https://github.com/indieweb/wordpress-indieweb/issues/68) based on user feedback from this weekend's IWS and open up the discussion more in a centralized place vs scattered through IRC discussions.

**Comment by miklb** [comment]
So another idea that came to me this evening. A whole new plugin, IndiePack if you will, that is that one click install all the things for someone who wants that and leave the IndieWeb plugin as is with an a la carte install option. Would just be a matter of making sure the two plugins spell out what they offer, link to each other, push the IndiePack to new users. The data stored as `mf2_*` is still the same if they decide to deactivate the IndiePack plugin and go a la carte after discovering they only want certain features.
This would allow the individual developers (and **cought** new devs **cough**) build and then just have the two options. I'm certainly game to being responsible for keeping up with the 2 different "master" plugins.

**Comment by snarfed** [comment]
@miklb glad to have you thinking and working on WordPress IndieWeb!

we've consistently had new WordPress users confused by the number of plugins and which ones they need. this was a big discussion topic at IWS2017 among @dshanske, @scottjenson, @cleverdevil, and more. they're definitely thinking along the same lines: make the bundle plugin one-click install and more like a wizard.

I'd strongly encourage us to do that inside the existing IndieWeb  plugin, and *not* in a new plugin that increases the proliferation confusion even more. otherwise, I'm all for it!

**Comment by miklb** [comment]
@snarfed I only proposed a new plugin as to take @pfefferle feedback into mind. I have as separate [issue open on the IndieWeb plugin](https://github.com/indieweb/wordpress-indieweb/issues/68), just seems the discussion clustered here.

I haven't put the time and energy into the plugins he and @dshanske have, and respect the idea that the plugins were bore out of dogfooding and scratching itches. Thus I'm hoping to help bridge the two perspectives.

**Comment by chrisaldrich** [comment]
<a href="https://github.com/indieweb/wordpress-uf2/issues/29#issuecomment-311222003" class="u-in-reply-to"></a>Using the JetPack "model" is a reasonable thought, though in this case the sub-plugins have a more cohesive bent rather than 20 things that have almost nothing to do with each other. Being able to leave the plugins as separate, stand-alone pieces (which can have the ability to inter-operate) is great, but there's also a need for a better integrated overall experience for Gen2+. 

Having watched the on-boarding experience people have had for the past two years with WordPress, the two biggest pieces I see are preventing redundancies/conflicts and having better theme support/choice. If the Indieweb plugin did _nothing_ but offer a better integration of the multitude of satellite plugins, it could work. I think the key is to have it do that and _only_ that. Everything else in it should be a standalone plugin. This is potentially the issue that had been seen before is that the IndieWeb plugin tried to do a few simple things AND bundle the others in. It's probably easier if everything is its own stand-alone that gets bundled into something bigger that helps with UI (like a JetPack-esque tabbed interface with some simple centralized settings).

Devs can pick and choose a plugin at a time (and continue to develop stand-alone pieces) while the masses can have a better integrated set up/workflow to on-board. 

@miklb, Thanks for taking on the idea of doing this!

**Comment by miklb** [comment]
I wonder then if those basic items in the IndieWeb plugin could be moved into the uf2 plugin. It's mostly `rel=me` tags and h-card right? That would fit into the scope of the plugin I think.

Then the IndieWeb plugin could purely be a on-boarding/bundling/auto-activating plugin for the gen2+ user.

**Comment by pfefferle** [comment]
hey miklb, as @snarfed already said, thanks a lot for beeing so active in the last days!

Sorry for maybe sounding a bit pessimistic, but I started the IndieWeb plugin for over 3 years ago and we tried a lot since then. One of the first ideas was to use a jetpack like style: https://github.com/indieweb/wordpress-indieweb/tree/94b5296a55cfce92eade18ca0725c76207cd7b58

The tagline even said

> The [IndieWeb](http://indiewebcamp.com/) version of WordPress' Jetpack plugin, it provides a bundle of the following plugins

The problem with that approach is, that plugins might be installed twice. If you bundle a subset of plugins, users often times do not understand which plugins are part of the bundle and which not. The find a lot of other IndieWeb plugins in the WP.org repo and might install them. So we ran into `Cannot redeclare class` errors so we had to add some checks https://github.com/willnorris/wordpress-hum/issues/16 to all of the plugins.

The next problem is versioning... It's hard to keep the IndieWeb plugin up to date if you have to keep up with the development of 10+ plugins.

So long story short, I would love to see a more user friendly maybe guided installation process and I like the idea of pre-installing some plugins. But I would highly recommend to not bundle them!

I also share the idea of @chrisaldrich to keep the IndieWeb plugin free of code that is not about installing plugins: https://github.com/indieweb/wordpress-indieweb/issues/40

**Comment by miklb** [comment]
Then my proposition is this:

- close this issue as `wontfix`
- create a milestone for 2.0 for the IW plugin
- begin moving the parts out of the IW plugin and merging them into the MF2 plugin
- Release a 2.0 of the IW plugin that is purely a guided installer.

I am in this for the long haul and will take on the responsibility of overseeing the upkeep of the 10+ plugins being up-to-date in the IW plugin.

I am not suggesting bundling plugins **IN** the IW plugin, rather auto-installing (if possible) and activating webmentons, semantic-linkbacks and syndication links. Plus any other plugins the community deems as must-have. Then helping users understand and activate the other IW related plugins.

**Comment by dshanske** [comment]
I am suggesting we table the issue of moving the identity parts out of the Indieweb plugin until we have the guided installer and/or other install enhancements. Let's see about enhancing the installer class we adopted(we can also submit our enhancements upstream separately), and then move from there.

**Comment by miklb** [comment]
why are they mutually exclusive? Work can be done in separate branches on both plugins.

**Comment by dshanske** [comment]
I am just saying that I would like to see some of the usability enhancements before we remove anything

**Comment by cleverdevil** [comment]
Just chiming in, since I was really impacted by my experience at IWS 2017 helping people get set up with IndieWeb sites on WordPress. The individual plugins themselves are awesome achievements that enable some really great features and capabilities for users, but currently:

1. They are named with and described by technical terms and jargon. Gen 2/3 users will not know what Webmention, Semantic Linkbacks, or Microformats2 are. Once they install the IndieWeb plugin, they are presented with a list of eight additional plugins to install, none of which have user-friendly names or descriptions.
2. Having to install a bunch of different plugins, activate them, and then configure them is a lot of work, and is very confusing. Instead, having some form of on-boarding wizard that interrogates the user about their needs/desires, and then installs/activates/configures that right plugins automatically would be much better.

I also agree that themes are an issue, but I think that its sort of a separate problem.

Over the weekend, one of my WordPress experts at work walked through the process of attempting to get set up with an IndieWeb WordPress site, and related to me his confusion and difficulty navigating the process. If he can't figure it out on his own, then our Gen 2 and Gen 3 users will have even less of a shot.

By the way, this feedback is in no way intended to diminish the amazing work that the community has done getting us to this point! Just a suggestion that an onboarding process inside of the IndieWeb plugin that enables Gen 2/3 users to get up and running as quickly as possible would be a massive help.

**Comment by miklb** [comment]
So I went ahead and checked with the plugin review team to make sure I knew what can & can't be done in regards to plugins requiring other plugins.

Short answer, you can't. 

You can recommend and walk a user through installing the plugins, but you can not auto-install & activate other plugins, even if you explicitly tell the user "hey, if you install and activate this plugin, it will install & activate these 3 other plugins."

I understand the concerns for security and what allowing that could open up.

We'll just need to design and architect the wizard accordingly.

**Comment by cleverdevil** [comment]
@miklb that's a shame!

I understand why this is the case, but it does present an obstacle. If the wizard itself can *install* other plugins on its own in response to the user allowing it, I think it'll be fine. As the user toggles the features / capabilities that they're interested in ("Support for rich comments and interactions from other websites," "Post likes, bookmarks, recipes, and more"), the appropriate plugins  ("Webmention," "Post Kinds") will be installed/enabled/configured.

If that's not the case, honestly, I see little benefit to having all of these plugins be separate. I think the user would be better served by using Jetpack as a model, and bundling all (or most) of the plugins and their capabilities into a single plugin to make the user experience better, allowing the user to install a single plugin to get their site IndieWeb-ified.

**Comment by miklb** [comment]
@cleverdevil correct, the user can click a button to install the plugin(s), but the plugin can not install without that user interaction.

To be clear, that is a rule for the wp.org repo. A plugin could be served from GitHub or somewhere else that a user could upload (either via SFTP or the admin interface) that could auto-install a bundle and then walk a user through other suggested plugins and configurations. There are methods to allow the plugin to be updated from a remote repo via the standard WordPress plugin update notifier. Meaning, once uploaded, wouldn't require uploading continued updates, assuming the install is configured to allow updates for other plugins (permissions, etc).

The discussion regarding bundling all the IW plugins at this point would require maintaining forks of all of the plugins, a more daunting task than just keeping up with an installer. 

This conversation has become a bit fractured, as @kraftbj [points out](https://github.com/indieweb/wordpress-uf2/issues/30#issuecomment-311676060) in the other thread, themes (and I assume plugins?) can throw dismissible warnings that for best experience X plugin(s) should be installed. But the theme must be fully functional (or at least not throw any php errors) without the added plugins.

**Comment by kraftbj** [comment]
One difficulty of multiple plugins is, for whatever reason, if someone updates one plugin, but not another, but Plugin A expects something from a newer version of Plugin B. It's possible with version checks and whatnot, but can lead to some odd user experiences. 

I think a lot of it depends on philosophy. Does Indieweb want a nice package that someone can just use to get the benefits (or a foundation set of benefits) of open/indieweb? If so, I would think that foundation should be all together and active by default. If there are other features/benefits that would continue to augment the experience, then there's the open question of how much of that should be optional but within the main plugin (a Jetpack option, so to speak) or how much of that should be an installer model like the current one.

I've said it somewhere else but my first experience of using webmentions (last week?) was activating the webmentions plugin first, setting up brid.gy, having ~1000 facebook/twitter responses come into my site without any semantic enhancement. The mentions looked horrible. I ended up deleting them, then later found the Indie Web plugin (and realized the webmention plugin I used and Indie Web were of the same group/community) and realizing that I needed Semantic Linkbacks plugin to make it worth using. I'm not sure how to force brid.gy to try to webmention everything again, so I gave up on backfilling. Not to sound stuck up, but I have commits (most of them worthless ;-) ) on every version of WordPress since like 3.6, worked on Jetpack for nearly four years, worked on WordPress.com VIP, and built countless sites for clients. If *I* had this kind of experience, I can only imagine we're losing folks who stumble across things in some way like I did and get the impression the technology is half-baked.

**Comment by cleverdevil** [comment]
You captured my sentiments exactly @kraftbj.

**Comment by pfefferle** [comment]
@kraftbj but you ignore Gen 1 users with this approach. How will you prevent Class collisions if some users install the single plugin together with the bundle? How will keep up with all the updates of all single plugins? How will you prevent collisions between different versions of self installed plugins that "overwrite" the bundled version?

**Comment by pfefferle** [comment]
As I said some comments above... This was exactly how the first version looked like and we had a lot of work/problems with that approach...

**Comment by snarfed** [comment]
background on "Gen 1": https://indieweb.org/generations

@pfefferle the indieweb wordpress world has been optimized for gen 1 for a while now, and as a result it's pretty inaccessible and hostile to gen 2 and below. worse, as @kraftbj demostrated it's even inaccessible to other gen 1 people who aren't us! i know you're not focused on gen 2+, but many of the rest of us are. if we can make it more accessible to gen 2 and below, i think that's worth a little inconvenience for gen 1.

...having said that, i don't think it will actually be worse for us. gen 1 developers can ignore the bundle plugin. if we want to dogfood individual plugins, we can still install the bundle and tell it not to install any specific plugins we've self installed. i don't think we have to choose between gen 1 friendly and gen 2+ friendly.

(for the webmention and semantic linkbacks plugins specifically, imho they should absolutely merge. @dshanske has agreed for a while and is working on that. i think the idea that you have to do new development in new plugins is a red herring, and often hurts usability for everyone.)

**Comment by pfefferle** [comment]
@snarfed your words in gods ear!

Regarding your Webmention/Semantic Linkbacks concerns... They are two plugins, because I hoped the Webmention part will be moving into core some times in the far future... I suspect, that the whole mf2 parsing stuff will not be "stable" any time soon, at least not that stable that it might be merged into core... That's it...

**Comment by cleverdevil** [comment]
The best way to get things into core is to get users. The best way to get users is to improve the user experience.

**Comment by pfefferle** [comment]
@cleverdevil might be true, but not in the last case... The semantic linkback plugin is not nearly in a solid state and uses libs that breaks WordPress' PHP compatibility...

**Comment by kraftbj** [comment]
A few ideas:

1. Auto-deactivate existing plugins when using the bundles that include the plugins now.
2. Coupled with 1, namespace the classes in the bundled plugin or
3. Only include the bundled plugins late (e.g. on the `plugins_loaded` hook) if the class doesn't exist (coupled with 1 or not). Kinda like pluggable.php in WP -- let individual plugins define the classes, but use the bundled ones if they're not defined. Granted, pluggable has plenty of issues too.
4. Use a PHP constant. A bit old school but: https://github.com/Automattic/pushpress/blob/master/pushpress.php#L17 In this case, WP.com has another class that is extending the origin pushpress class. We set this constant to the other class instead so pushpress lives on wp.com as it does in the .org repo, but we can still load our own version (for things like moving the pings to our separate async jobs system, handling some special one-off VIP sites, etc).

For the Webmention/Semantic Linkbacks, we can keep the codebase separated within the one plugin so it's easy to pull out one for core inclusion. The main plugin file includes two other files, one for each plugin. It's a bit more hefty, but if that's the blocker...

**Comment by miklb** [comment]
I still stand by [my proposal yesterday](https://github.com/indieweb/wordpress-uf2/issues/29#issuecomment-311363720)

> Then my proposition is this:
> 
> close this issue as wontfix
> create a milestone for 2.0 for the IW plugin
> begin moving the parts out of the IW plugin and merging them into the MF2 plugin
> Release a 2.0 of the IW plugin that is purely a guided installer.

This would allow us to seize on some of the momentum coming out of IWS and continue the conversation of how to make the guided installer better, or, package the plugins into one consumer/gen2+ specific plugin. Seems the discussion in #30 and interoperability needs further fleshing out, and that conversation is being diluted by touching on the ideas of a master plugin.

I sincerely appreciate the thought and work that's gotten IndieWeb this far.

@anomalily's [keynote was posted today](https://www.youtube.com/watch?list=PLk3TtIJ31hqrLIPqz55TczawWu-30cnXM&v=1VGX8iBWrTE) and that experience and growing the community is my motivation at this point.

**Comment by pfefferle** [comment]
Perhaps I am really not the right one to lead the discussion... I think it's worse a try... I am curious about the V2...

@miklb sure, but I would start a new issue at the IndieWeb plugin and I would reference this discussion.

**Comment by miklb** [comment]
@pfefferle  v2 would roughly mean that the IW plugin is **just** a guided installer and introduction to IndieWeb and WordPress. Any functionality is moved to mf2 plugin. We could get input from some UX designers on the wizard/guided installer. I'll flesh that out and make a new issue later this evening.
